Remove ancestor

gchoqueux opened the pull request:
### Refactor
**Remove node's ancestor**
The ancestor gave the textures downscaled WMTS and WMS. Now, ancestor is replaced by direct parent. This replacement is effective for color and elevation.

peppsac commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -109,22 +109,23 @@ WMTS_Provider.prototype.url = function url(coWMTS, layer) {
     }
 
-    var url = this.url(cooWMTS, layer);
+    var url = this.url(coordWMTS, layer);
```
When `searchInParent` is true you could skip the `this.url()` call and directly use parent's texture.

If this is done on all providers, we could remove their usage of `CacheResources` which IMHO is a good idea:
* the browser already cache network request
* we never evict data from this cache

This would allow to effectively delete unused THREE.Texture which is currently not done (even after calling material.dispose() because CacheResource still own a reference).

gchoqueux replied:
Yes, this is the direction I want to take. 
I can do the job now?

`material.dispose()` frees VRAM and not RAM (CacheResource have reference in RAM)

peppsac replied:
Yes, it would be nice if you have the time to implement it.
Also you may want to squash the 3 `remove ancestor in loading` commits.

gchoqueux replied:
CacheRessource is necessary for neighboring PM textures

gchoqueux commented:
updated

peppsac commented on index.html:
```diff
@@ -53,9 +53,9 @@
-            urlImageryLayers.push('examples/layers/JSONLayers/OrthosCRS.json');
-            urlImageryLayers.push('examples/layers/JSONLayers/ScanEX.json');
-            urlImageryLayers.push('examples/layers/JSONLayers/Region.json');
+            // urlImageryLayers.push('examples/layers/JSONLayers/OrthosCRS.json');
```
Why did you remove those?
Also your commit title misses a "commit_type:" prefix

gchoqueux replied:
The external returns on the demo are:
  * Loading speed (due to the 4 layers)
  * Unnecessary layers: region (ugly), CRS (duplicate with ortho) and scan (limited to France)

peppsac replied:
Ok (just copy-paste your comment in the commit message)

peppsac commented on src/Core/Commander/Providers/TileProvider.js:
```diff
@@ -44,6 +47,7 @@ TileProvider.prototype.executeCommand = function executeCommand(command) {
         bbox,
         level: (command.level === undefined) ? (parent.level + 1) : command.level,
         segment: 16,
+        light: command.layer.light,
```
I don't think `TileProvider` needs to be aware of lighting.
Maybe you could move lights handling in NodeProcess?

gchoqueux replied:
completely agree

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -22,22 +18,11 @@ import { UNIT } from 'Core/Geographic/GeoCoordinate';
-        return texture;
-    };
+function WMS_Provider(options) {
+    WMTS_Provider.call(this, options);
```
Instead of inherinting from `WMTS_Provider`, maybe we could build a helper class for all those protocols (WFS, WMS, WMTS)?

gchoqueux replied:
yes, I also thought about a imageTiledProvider  to provide image tiled for WMTS, WMS or big texture.

peppsac replied:
I'd rather not introduce this inheritance.
So I belive you need to either:
- introduce a `OGCWebServiceHelper.js` file holding helper methods (see `Fetcher.js` for a good example)
- live a few more weeks with some duplicate codes between `WMTS` and `WMS`, until someone is willing to refactor this code.

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -73,6 +58,7 @@ WMS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer)
     layer.style = layer.style || '';
     layer.transparent = layer.transparent || false;
     layer.bbox = layer.bbox || new BoundingBox();
+    layer.options = { tileMatrixSet: 'WGS84G' };
```
Why this change?

gchoqueux replied:
Because I need to know how the wms is tiled

peppsac replied:
But: 
- they might not be tiled like this. For instance the Lyon opendata wms server is in meter and you can query any bbox inside its extents so the `tileMatrixSet` makes no sense in this case.
- if you add an option use the `layer.foo = layer.foo || my_default_option`

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -95,125 +81,57 @@ WMS_Provider.prototype.getColorTexture = function getColorTexture(tile, layer, b
-            ancestor.bbox :
-            tile.bbox;
-
+        const searchInParent = tile.materials[0].getColorLayerLevelById(layer.id) < 0 && tile.parent.materials[0].getColorLayerLevelById(layer.id) > -1;
```
No, you can't depend on `tile.materials[0].getColorLayerLevelById(layer.id)` to do this.
Actually you can't depend on `LayeredMaterial` at all.
Add the needed params to the `command` instead (for instance the parent's texture for this layer).

gchoqueux replied:
ok

peppsac commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -104,62 +114,69 @@ WMTS_Provider.prototype.url = function url(coWMTS, layer) {
+    const searchInParent = tile.material.getElevationLayerLevel() < 0;
+
+    if (searchInParent || coordWMTS.zoom > layer.zoom.max) {
+        const parentZoom = tile.parent.material.getElevationLayerLevel();
```
idem: I'd rather not introduce a dependency to `LayeredMaterial` here.

peppsac commented on src/Globe/Atmosphere.js:
```diff
@@ -96,95 +93,28 @@ function Atmosphere(ellipsoid) {
-            type: 'f',
-            value: 1.0,
-        },
+        v3LightPosition: { value: LIGHTING_POSITION.clone().normalize() },
```
Nice change, but it would have been easier to review if you had split all the code cleanup in a separate PR.

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -420,4 +419,20 @@ LayeredMaterial.prototype.getLayerLevel = function getLayerLevel(layerType, laye
+        const textureIndex = this.getTextureOffsetByLayerIndex(index);
+        return this.textures[l_COLOR].slice(textureIndex, textureIndex + count);
+    } else {
+        return [];
```
```js
throw new Error(`Requesting textures for invalid layer : ${layer}`);
```
Also replace function param name : `layer` -> `layerId`

gchoqueux replied:
ok

peppsac commented on src/Renderer/c3DEngine.js:
```diff
@@ -28,6 +28,7 @@ function c3DEngine(scene, positionCamera, viewerDiv, debugMode, gLDebug) {
     this.viewerDiv = viewerDiv;
     this.debug = debugMode;
     this.scene3D = new THREE.Scene();
+    this.scene3D.autoUpdate = false;
```
This change is unrelated to the current PR so I'd prefer that you submit a separate PR for this.

peppsac commented on src/Scene/LayersConfiguration.js:
```diff
@@ -4,6 +4,7 @@
  * Description: Le layer est une couche de données. Cette couche peut etre des images ou de l'information 3D. Les requètes de cette couche sont acheminées par une interfaceCommander.
  *
  */
+import { l_COLOR, l_ELEVATION } from 'Renderer/LayeredMaterial';
```
I'd rather not do this.
If you really need to be able to do `layer.type = l_ELEVATION;` then the `l_ELEVATION` constant probably belong to this file.
But I'd rather not have this at all - and it's unrelated to the current PR scope anyway ?

peppsac reviewed (changes requested):
The new mechanism works well, but I left some general architecture comments.

peppsac commented:
Btw, you could have done 4 different PRs, 1 for each "theme":

* fix (Loading texture) : error loading elevation #236 
* fix (Layer) : error in option layer for lighting
* removes layers for demo Front page
* and the 2 providers refactoring commit

They would have been easier to review and faster to merge (hint: it's not too late to do this :))

(yes my own current PR has 17 commits ranging from layer refactoring to typo fixing but... do as I say not as I do :))

peppsac commented:
Looks better with the last commit. Can you split this PR in 4 PR (see comment https://github.com/iTowns/itowns2/pull/244#issuecomment-279326570) so I can review/merge ?

gchoqueux committed:
refactor ( provider) : remove ancestor in WMTS and WMS provider

Now we fetch textures directly in parent

gchoqueux committed:
refactor ( provider) : Factorization WMTS and WMS provider

gchoqueux committed:
refactor (provider) : pass parent's texture in command

remove layer type in LayerConfiguration

gchoqueux commented:
rebased

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -73,6 +58,9 @@ WMS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer)
     layer.version = layer.version || '1.3.0';
     layer.style = layer.style || '';
     layer.transparent = layer.transparent || false;
+    layer.bbox = layer.bbox || new BoundingBox();
```
Empty `BoundingBox` can't be built anymore.
IMHO a WMS layer requires a boundingbox to work correctly so we'd better throw an error if it's missing.
Of course we'll be able to improve this when `GetCapabilities` is implemented :)

peppsac replied:
Of course the same goes for `projection` attribute: it's mandatory.

gchoqueux replied:
projection?

gchoqueux replied:
ah ok  the CRS

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -90,130 +78,60 @@ WMS_Provider.prototype.tileInsideLimit = function tileInsideLimit(tile, layer) {
-    if (result.texture !== undefined) {
-        return Promise.resolve(result);
-    }
+    const url = this.url(tile.bbox.as('EPSG:4326'), layer);
```
`tile.bbox.as(layer.projection)`

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -90,130 +78,60 @@ WMS_Provider.prototype.tileInsideLimit = function tileInsideLimit(tile, layer) {
-        result.texture.needsUpdate = true;
+    return this.getColorTextureByUrl(url).then((texture) => {
+        result.texture = texture;
+        result.texture.coordWMTS = tile.wmtsCoords[layer.options.tileMatrixSet][0];
```
Why is this needed? WMS has nothing to do with `coordWMTS` right?

gchoqueux replied:
WMS is tiled and works like WMTS.
I need to know his correspondence with tile coordinate

Jeremy-Gaillard replied:
WMS is not a tiled protocol. If you need a coordWMTS attribute, I suggest moving the assignment to NodeProcess or another place which works exclusively with tiles.

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -90,130 +78,60 @@ WMS_Provider.prototype.tileInsideLimit = function tileInsideLimit(tile, layer) {
+WMS_Provider.prototype.getXbilTexture = function getXbilTexture(tile, layer) {
+    const url = this.url(tile.bbox, layer);
+    return this.getXBilTextureByUrl(url, new THREE.Vector3(0, 0, 1)).then((result) => {
+        result.texture.coordWMTS = tile.wmtsCoords[layer.options.tileMatrixSet][0];
```
Same, `coordWMTS` is irrelevant here.

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -90,130 +78,60 @@ WMS_Provider.prototype.tileInsideLimit = function tileInsideLimit(tile, layer) {
+    const layer = command.layer;
 
-    var supportedFormats = {
+    computeTileWMTSCoordinates(tile, layer, this.projection);
```
Unneeded.

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -398,16 +337,15 @@ function updateNodeElevation(scene, quadtree, node, layersConfig, force) {
     if (bestLayer !== null) {
         node.layerUpdateState[bestLayer.id].newTry();
 
+        const searchInParent = (!node.isElevationLayerLoaded() && node.parent.isElevationLayerLoaded()) || (node.level > bestLayer.zoom.max);
```
Can you add a comment explaining why the test is different to the color one?

gchoqueux replied:
Because elevation is composed of 2 layers

peppsac replied:
I meant a comment in the *code*, not the PR.

gchoqueux commented:
I need WMTS coordinate  to search in parent texture

peppsac commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,129 @@
+const getTextureFloat = function getTextureFloat(buffer) {
+    // Start float to RGBA uint8
+    // var bufferUint = new Uint8Array(buffer.buffer);
+    // var texture = new THREE.DataTexture(bufferUint, 256, 256);
```
Maybe remove this commented code

peppsac commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -73,6 +58,8 @@ WMS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer)
     layer.style = layer.style || '';
     layer.transparent = layer.transparent || false;
+    layer.options = {};
+    layer.options.tileMatrixSet = layer.tileMatrixSet || 'WGS84G';
```
See previous comment thread https://github.com/iTowns/itowns2/pull/244#discussion_r104377515
(TL;DR: *WMS is not a tiled protocol and has nothing to do with WMTS coords*)

peppsac commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -4,45 +4,13 @@
-            // Start float to RGBA uint8
-            // var bufferUint = new Uint8Array(buffer.buffer);
-            // var texture = new THREE.DataTexture(bufferUint, 256, 256);
+import Tish from './TiledImageTools';
```
`Tish` ??

peppsac commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -109,147 +73,65 @@ WMTS_Provider.prototype.url = function url(coWMTS, layer) {
+    const pitch = new THREE.Vector3(0.0, 0.0, 1.0);
+
+    if (parentTextures) {
+        coordWMTS = Tish.WMTS_WGS84Parent(
```
This only works with `WGS84G` tileMatrixSet right? (As you assume there's only one texture)
If this is right, add throw an error when this condition is false.

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -394,26 +394,41 @@ LayeredMaterial.prototype.isLayerTypeDownscaled = function isLayerTypeDownscaled
+        const textureIndex = this.getTextureOffsetByLayerIndex(index);
+        return this.textures[l_COLOR].slice(textureIndex, textureIndex + count);
+    } else {
+        throw new Error('No known layer :', layerId);
```
```js
throw new Error(`Invalid layer id "${layerId}"`);
```

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -253,48 +236,25 @@ function updateNodeImagery(scene, quadtree, node, layersConfig, force) {
-        if (currentLevel === -1) {
-            ancestor = findAncestorWithValidTextureForLayer(node, l_COLOR, layer);
-        } else {
-            var targetLevel = chooseNextLevelToFetch(layer.updateStrategy.type, node.level, currentLevel, layer.updateStrategy.options);
```
You're dropping all network strategy code.
Why?

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -253,48 +236,25 @@ function updateNodeImagery(scene, quadtree, node, layersConfig, force) {
-            /* specific params */
-            ancestor,
-            /* redraw only if we're aren't using a texture from our parent */
-            redraw: (ancestor == null),
```
Why do you remove `redraw` param ?

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -253,48 +236,25 @@ function updateNodeImagery(scene, quadtree, node, layersConfig, force) {
                 } else if (result.texture) {
-                    result.texture.level = level;
+                    if (!result.texture.coordWMTS) {
+                        result.texture.coordWMTS = node.wmtsCoords[layer.options.tileMatrixSet][0];
```
Eeek. If this is mandatory this can be handled by the provider.

gchoqueux replied:
[except for WMS, see comment](https://github.com/iTowns/itowns2/pull/244#discussion_r104423818)

gchoqueux committed:
refactor (WMTS/WMS) : add tiledImageTools

Factorize common functions in tiledImageTools

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -199,6 +199,10 @@ ApiGlobe.prototype.removeImageryLayer = function removeImageryLayer(id) {
 
 ApiGlobe.prototype.addElevationLayer = function addElevationLayer(layer) {
+    if (layer.protocol === 'wmts' && layer.options.tileMatrixSet !== 'WGS84G') {
+        throw new Error('Elevation layer with wmts protocol support only WGS84G projection');
```
If I understand correctly, this is a current limitation of itowns (well, in its globe mode at least). To make this clearer, I suggest:

> Only WSG84 tileMatrixSet is currently supported for WMTS elevation layers

(If I'm not mistaken, WSG84 is a geometry/tile matrix set, not a projection)

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+
+export default {
+    cropXbilTexture(texture, pitch) {
+        return _cropXbilTexture(texture, pitch);
```
This proxy is not needed, you can rename `_cropXbilTexture` in `cropXbilTexture` and expose this directly,
```js
export default {
    cropXbilTexture,
    getColorTextureByUrl(url) {
          /* ... */  
    }
    /* ... */
```

Btw, defining named function and exposing them in default object instead of using the property shortcut allows you to reference them directly if they need to call each other. Otherwise, a method B would need to use  `this.methodA`. Referencing them directly is good, because it allows functions to be passed around as ref without worrying about `this`, so avoiding the need to `bind`. 

To be clearer, here in this particular case:

```js 
function a() { /*...*/ }
function b() { /* ... */}
export default {
    a,
    b,
}
```

is better than 
```js
export default {
    a() { /*...*/ },
    b() {/* ...*/ }
}
```
because in the first case, b can call a() even if you take it out of the object (by doing `let myB = obj.b`). In the second case, you need to use `this.a()` in b, therefore you need to use `bind` if you want to pass this method around. As `TiledImageTools` is a helper (has no instances) the first pattern can be used easily and is more robust

autra replied:
But actually, I think we don't need this method anyway.

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+import CoordWMTS from '../../Geographic/CoordWMTS';
+import { UNIT } from '../../Geographic/Coordinates';
+
+const SIZE_TEXTURE_TILE = 256;
```
If I'm not mistaken SIZE_TEXTURE_TILE is dependent of the TileMatrixSet. I know that the only TileMatrixSets we currently support use this value, but I'd rather not make things worse here, and use the parametized values in functions that uses this instead (layer.width and layer.height). It would allow to support other TileMatrixSets more easily.

gchoqueux replied:
yes, need new PR

autra replied:
Ok, it's coming from old code. I might change this in the `getCapabilities` PR.

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+    getColorTextureByUrl(url) {
+        const textureCached = cache.getRessource(url);
+
+        if (textureCached !== undefined) {
```
`if (textureCached)` ?

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+        });
+    },
+    getXBilTextureByUrl(url, pitch) {
+        const textureCache = cache.getRessource(url);
```
`const cachedTexture`

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+    getXBilTextureByUrl(url, pitch) {
+        const textureCache = cache.getRessource(url);
+
+        if (textureCache !== undefined) {
```
`if (cachedTexture)`

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+        }
+
+        return ioDXBIL.read(url).then((result) => {
+            // RGBA is needed for navigator with no support in texture float
```
Is this a TODO, a FIXME or a general comment explaining something?

autra replied:
(if it's a TODO, as I suspect it is: 
```js
// TODO RGBA texture must be used for browsers without float texture support
```
)

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+        return ioDXBIL.read(url).then((result) => {
+            // RGBA is needed for navigator with no support in texture float
+            // In RGBA elevation texture LinearFilter give some errors with nodata value.
+            // need to rewrite sample function in shader
```
Is this a TODO? Should be in the said shader then?

That being said, I know this comes from `WMTS_Provider`, so I'm not requiring these comments to be fixed here (it's nice if you do though).

gchoqueux replied:
The shader `GlobeFS.glsl` already takes into account the comment

autra commented on src/Scene/NodeProcess.js:
```diff
@@ -99,6 +100,7 @@ NodeProcess.prototype.subdivideNode = function subdivideNode(node, camera, param
                 // update Imagery wmts
                 for (const layer of colorLayers) {
                     if (layer.tileInsideLimit(child, layer)) {
+                        TiledImageTools.computeTileWMTSCoordinates(child, layer);
```
This is only for WMTS layer isn't it? if so, why are you doing this here?

autra replied:
@Jeremy-Gaillard as you requested this change, any objection to have this in `WMTS_Provider`, for example in `executeCommand`?

gchoqueux replied:
I need WMTS coordinate here because the **WMS** is also required in tile

gchoqueux replied:
This is done here because we need to know the WMTS coordinates to know if we require new texture or not

autra replied:
Ok I see. WMTS is a protocol. You're speaking about our *TileMatrixSet*. Let's call that `computeTileMatrixSetCoordinates`. Same for the other occurrences.

autra replied:
I do hope we will support other tileMatrixSets than WSG84G in the future.

gchoqueux replied:
You really want to change? Because wmts coordinate is not a protocol.
It already supports PM

autra replied:
[WMTS *is* a protocol](https://en.wikipedia.org/wiki/Web_Map_Tile_Service). 

Yes, we support WSG84G TileMatrixSet + we kind of support PM TileMatrixSet. But actually, everything fits together because our geometry follow the same tile scheme. It is specific to the globe, and even more : to a IGN-homemade TileMatrixSet. So this still needs renaming, but I'm gonna start a new discussion in the right place (where this method is defined) proposing something in my next review.

autra commented on src/Scene/NodeProcess.js:
```diff
@@ -119,6 +121,7 @@ NodeProcess.prototype.subdivideNode = function subdivideNode(node, camera, param
                 const elevationLayers = params.layersConfig.getElevationLayers();
                 let canHaveElevation = false;
                 for (const layer of elevationLayers) {
+                    TiledImageTools.computeTileWMTSCoordinates(child, layer);
```
Ok, we only support wmts elevation, but still, I still feel this doesn't belong here.

gchoqueux replied:
Same 
> This is done here because we need to know the WMTS coordinates to know if we require new texture or not

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+                projection.getCoordWMTS_WGS84(tileCoord, tile.bbox, tileMatrixSet);
+        }
+    },
+    WMTS_WGS84Parent(cWMTS, levelParent, pitch) {
```
Should probably stay in `WMTS_Provider`.

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+
+        return new CoordWMTS(levelParent, r, c);
+    },
+    WMS_WGS84Parent(bbox, bboxParent) {
```
Should probably stay in `WMS_Provider`.

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+            return result;
+        });
+    },
+    computeTileWMTSCoordinates(tile, wmtsLayer) {
```
See my comments in NodeProcess. The way we solve this discussion will determine the location of this method.

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+
+const SIZE_TEXTURE_TILE = 256;
+
+// CacheRessource is necessary for neighboring PM textures
```
Can you elaborate?

Is it because you need to get several PM texture to cover a WSG84 tile?

gchoqueux replied:
The PM textures overlap several tiles WGS84, it is to avoid net requests

autra replied:
Ok then.

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+const SIZE_TEXTURE_TILE = 256;
+
+// CacheRessource is necessary for neighboring PM textures
+// Info : THREE.js have cache image https://github.com/mrdoob/three.js/blob/master/src/loaders/ImageLoader.js#L25
```
Yes, and unfortunately, if I'm not mistaken, they fail to evict it too :-(

gchoqueux replied:
that is to say?

autra replied:
THREE.js uses cache, but never evict it, never check its validity, doesn't have any limits in size. It's really bad. I'm just commenting on the `Info` comment. No action needed in this PR :-)

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+
+        return new THREE.Vector3(x, y, scale);
+    },
+};
```
So this factorizes code from both `WMS_Provider` and `WMS_Provider`. It cannot be called `TiledSomething`.

@peppsac 's proposal (`OGCWebServiceHelper`) sounds good to me. Feel free to propose something else though.

autra commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -20,23 +16,9 @@ import BoundingBox from '../../../Scene/BoundingBox';
 
-WMS_Provider.prototype = Object.create(Provider.prototype);
-
 WMS_Provider.prototype.constructor = WMS_Provider;
```
This line is now unneeded.

autra commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -90,130 +76,52 @@ WMS_Provider.prototype.tileInsideLimit = function tileInsideLimit(tile, layer) {
+        return func(tile, layer);
     } else {
         return Promise.reject(new Error(`Unsupported mimetype ${layer.format}`));
     }
```
Refacto proposal if you want, from line 111:

```js
switch (layer.format) {
    case 'image/png':
    case 'image/jpg':
    case 'image/jpeg':
        return this.getColorTexture(tile, layer);
    case 'image/x-bil;bits=32':
        return this.getXbilTexture(tile, layer);
    default:
        return Promise.reject(new Error(...));
```

peppsac replied:
Note that the elevation texture special case handling doesn't really belong to this provider.
For instance, I might have a WMS server supplying grayscale image to use as a MNT (and I've done this at least twice for demos built with itowns).

So deducing image usage from image format is wrong. It does'nt require fixing here, but keep this in mind while refactoring.

autra commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -4,45 +4,13 @@
 
-WMTS_Provider.prototype = Object.create(Provider.prototype);
-
 WMTS_Provider.prototype.constructor = WMTS_Provider;
```
Not needed any more

autra commented on src/Globe/TileMesh.js:
```diff
@@ -159,19 +159,34 @@ TileMesh.prototype.setBBoxZ = function setBBoxZ(min, max) {
 
-TileMesh.prototype.isColorLayerDownscaled = function isColorLayerDownscaled(layer) {
+TileMesh.prototype.getLayerTextures = function getLayerTextures(layerType, layerId) {
     var mat = this.materials[RendererConstant.FINAL];
```
Proposal: in this method an the three below: use `const mat`, or even better, get rid of the intermediary var. The engine will optimize it away anyway, but it would be faster when correcting linter errors.

autra commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -362,29 +362,29 @@ LayeredMaterial.prototype.getLoadedTexturesCount = function getLoadedTexturesCou
+        return this.textures[l_COLOR][this.getLayerTextureOffset(layerId)].coordWMTS.zoom < level;
+    } else {
+        return false;
+    }
```
nits
```js
return this.textures[l_COLOR][this.getLayerTextureOffset(layerId)] &&
    this.textures[l_COLOR][this.getLayerTextureOffset(layerId)].coordWMTS.zoom < level;
```
```

autra commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -362,29 +362,29 @@ LayeredMaterial.prototype.getLoadedTexturesCount = function getLoadedTexturesCou
-            if (this.textures[l_COLOR][offset].level < level) {
+            const texture = this.textures[l_COLOR][offset];
+            if (texture.coordWMTS.zoom < level) {
                 return true;
```
nits : 
```js
return this.textures[l_color][offset].coordWMTS.zoom < level;
```

gchoqueux replied:
I don't think because we must check all the textures

autra replied:
Yes, sorry I meant
```js
const texture = this.textures[l_COLOR][offset];
return textures.level < level && textures.coordWMTS.zoom < level;
```
`if (condition) return true` is always equivalent to `return condition`.

btw why do we need to test textures.level *and* textures.coordWMTS.level ? Isn't it redundant ?

gchoqueux replied:
we test only :`texture.coordWMTS.zoom < level`

autra replied:
Ok ok yes, I miread the diff in my second comment. I thought that's what you meant by 

> I don't think because we must check all the textures

So we're back at my first comment, which is actually correct: 

> nits :
> ```js
> return this.textures[l_color][offset].coordWMTS.zoom < level;
> ```

And 

> `if (condition) return true` is always equivalent to `return condition;`

Do you see what I mean?

autra replied:
Ok, the return false is after the loop, my mistake. Sorry for the noise :-)

autra commented on src/Scene/NodeProcess.js:
```diff
@@ -253,48 +237,36 @@ function updateNodeImagery(scene, quadtree, node, layersConfig, force) {
 
-        let ancestor = null;
-
+        node.layerUpdateState[layer.id].newTry();
```
There is still a case where we `continue` below, can you explain why you mark a new try here and not after this `continue`?

autra commented on src/Scene/NodeProcess.js:
```diff
@@ -253,48 +237,36 @@ function updateNodeImagery(scene, quadtree, node, layersConfig, force) {
-                    result.texture.level = level;
+                    if (!result.texture.coordWMTS) {
+                        result.texture.coordWMTS = node.wmtsCoords[layer.options.tileMatrixSet || 'WGS84G'][0];
+                    }
```
Maybe this belongs to `WMTS_Provider.executeCommand`

gchoqueux replied:
I need this for WMS too. because WMS is requested by tile

autra replied:
ok

autra commented on src/Scene/NodeProcess.js:
```diff
@@ -398,16 +354,18 @@ function updateNodeElevation(scene, quadtree, node, layersConfig, force) {
         node.layerUpdateState[bestLayer.id].newTry();
 
+        // Elevation layer search in parent, from the moment it exceeds its maximum zoom
+        const searchInParent = (!node.isElevationLayerLoaded() && node.parent.isElevationLayerLoaded()) || (node.level > bestLayer.zoom.max);
```
let's invert the shortcut here and test `node.level > bestLayer.zoom.max` first. It's more readable and would perform better, because isElevationLayer can potentially access some pretty big indices on array. I don't know if it is something noticeable here, but inverting this condition is cheap so...

autra commented on src/Scene/NodeProcess.js:
```diff
@@ -418,13 +376,9 @@ function updateNodeElevation(scene, quadtree, node, layersConfig, force) {
-                    terrain.min = (ancestor || node).bbox.bottom();
-                    terrain.max = (ancestor || node).bbox.top();
+                    terrain.min = (node.parent || node).bbox.bottom();
+                    terrain.max = (node.parent || node).bbox.top();
```
shouldn't it be `(node || node.parent)`?

autra commented on src/Core/Commander/Providers/TiledImageTools.js:
```diff
@@ -0,0 +1,126 @@
+const ioDXBIL = new IoDriver_XBIL();
+const projection = new Projection();
+
+const _cropXbilTexture = function _cropXbilTexture(texture, pitch) {
```
Actually, I don't see the point of this method: it only calls `computeMinMaxElevation` and returns an object containing unchanged results from this method, and *unchanged* parameters. It makes the API of this file unclear.

autra replied:
let's call `computeMinMaxElevation direction and in providers

```js
const {min, max} = ioDXBIl.computeMinMax(/* ... */);
return Promise.resolve({pitch, texture, min, max});
```

gchoqueux replied:
Ok, I took note of the remark, but i factorize in `TiledImageTools.js `

autra replied:
Why? It makes no sense to do that. `TiledImageTools` is not a promise factory. Factorization is only good when it's useful. And it's not useful to avoid a `Promise.resolve`. Adding a method just for that contribute to make the code harder to read and to maintain.

gchoqueux replied:
I will look at

autra commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -109,147 +73,65 @@ WMTS_Provider.prototype.url = function url(coWMTS, layer) {
     } else {
         return Promise.reject(new Error(`Unsupported mimetype ${layer.options.mimetype}`));
     }
 };
```
See refacto proposal in `WMS_Provider`

autra commented on src/Scene/NodeProcess.js:
```diff
@@ -253,48 +237,36 @@ function updateNodeImagery(scene, quadtree, node, layersConfig, force) {
             earlyDropFunction: refinementCommandCancellationFn,
-            /* specific params */
-            ancestor,
+            parentTextures: searchInParent ? node.parent.getLayerTextures(l_COLOR, layer.id) : undefined,
```
When we search in `parentTextures`, things are completely synchronous. So we don't need to pass through the scheduler. Not doing so will have several advantages:

- it would be clearer to the reader that things are synchrone if it is another method call.
- it would be easier to go multi-thread with Workers. If we go multithread, we would have no guarantee the parentTexture will be split before children receives the requested new tiles.
- It would make `provider.executeCommand` simpler.

autra replied:
Basically, instead of computing a `searchInParent` boolean, let's do the work there, and avoid the command scheduling in this case.

Scheduling is only useful when we have some network operations to do (because then we want to parallelize things, to wait for completion, to handle errors etc...)

gchoqueux replied:
Yes it is true, it should be seen in a new PR

We must not lose sight of the prioritization of the requests with the scheduler.

autra replied:
> it should be seen in a new PR

This is something you add in *this* PR. I understand we delay refactoring of existing code, but here, this is not the case. Moreover, it seems to me that the goal of this PR was to remove ancestor, in order to remove the transitional state where a node has children, but they are not ready yet. If we achieve this original goal, this would greatly simplify the code and perform better. For this, this children preparation must be synchronous. 

Yes the scheduler prioritize things, but here, there is no notion of priority : it should be done now. Also, this scheduling has runtime cost (using an external lib, sorting the queue...). We should use it only when we must.

autra reviewed (changes requested):
Things are mostly looking good :-) There is a lot of comments, but the PR is big, so...

gchoqueux commented:
thanks for reviewing @autra 
PR updated

autra commented on src/Scene/NodeProcess.js:
```diff
@@ -355,7 +355,7 @@ function updateNodeElevation(scene, quadtree, node, layersConfig, force) {
 
         // Elevation layer search in parent, from the moment it exceeds its maximum zoom
-        const searchInParent = (!node.isElevationLayerLoaded() && node.parent.isElevationLayerLoaded()) || (node.level > bestLayer.zoom.max);
+        const searchInParent = (!node.isElevationLayerLoaded() && node.parent.isElevationLayerLoaded()) || (bestLayer.zoom.max < node.level);
```
No I meant:
```js
const searchInParent = (bestLayer.zoom.max < node.level) || (!node.isElevationLayerLoaded() && node.parent.isElevationLayerLoaded());

autra replied:
This way, node.isElevationLayerLoaded is not called if first condition is true.

autra commented:
> PR updated

Still some actions needed
- https://github.com/iTowns/itowns2/pull/244#discussion_r111219933
- https://github.com/iTowns/itowns2/pull/244#discussion_r111165042

And I have added some remarks and answers.

gchoqueux commented:
new update

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -199,6 +199,10 @@ ApiGlobe.prototype.removeImageryLayer = function removeImageryLayer(id) {
 
 ApiGlobe.prototype.addElevationLayer = function addElevationLayer(layer) {
+    if (layer.protocol === 'wmts' && layer.options.tileMatrixSet !== 'WGS84G') {
+        throw new Error('Only WSG84 tileMatrixSet is currently supported for WMTS elevation layers');
```
`Only WSG84G...` sorry I might have mispelled it in my previous review.

autra commented on src/Core/Commander/Providers/OGCWebServiceHelper.js:
```diff
@@ -0,0 +1,84 @@
+            return result;
+        });
+    },
+    computeTileMatrixSetCoordinates(tile, wmtsLayer) {
```
I'm not requiring anything here as it is a big topic, just sharing some thoughts:
- all these wmtsFoo variables and methods should get renamed to tileMatrixSetFoo
- In the future `WGS84ToWMTS` might be a `coordinate.toTileMatrixSetCoordinates( tileMatrixSetId)` or a `bbox.toTileMatrixSetCoordinates(tileMatrixSetId` (knowing that a coordinate embed its crs). This would allow us to support custom tileMatrixSet.

autra commented on src/Core/Commander/Providers/WMS_Provider.js:
```diff
@@ -73,6 +74,7 @@ WMS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer)
     layer.version = layer.version || '1.3.0';
     layer.style = layer.style || '';
     layer.transparent = layer.transparent || false;
+    layer.options = {};
```
Are you sure about that? layer.options might not be empty (ex `Region.json`)

autra commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -109,147 +86,69 @@ WMTS_Provider.prototype.url = function url(coWMTS, layer) {
+    const pitch = new THREE.Vector3(0.0, 0.0, 1.0);
+
+    if (parentTextures) {
+        coordWMTS = WMTS_WGS84Parent(
```
If I'm not mistaken, `WMTS_WGS84Parent` returns a new object, but here you're not doing anything with `coordWMTS` before returning. Likely a mistake ?

autra commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -394,26 +391,41 @@ LayeredMaterial.prototype.isLayerTypeDownscaled = function isLayerTypeDownscaled
 };
 
-LayeredMaterial.prototype.getLayerLevel = function getLayerLevel(layerType, layer) {
+LayeredMaterial.prototype.getLayerLevel = function getLayerLevel(layerType, layerId) {
```
If I'm not mistaken, you removed the only calls to this method, it's now unused. Let's remove it.

autra replied:
The switch is now entirely in NodeProcess I guess.

autra commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -394,26 +391,41 @@ LayeredMaterial.prototype.isLayerTypeDownscaled = function isLayerTypeDownscaled
+    const index = this.indexOfColorLayer(colorLayerId);
     if (index === -1) {
-        index = 0;
+        return -1;
```
```js
throw new Error(`Invalid layer id ${colorLayerId}`)
``` 
?

gchoqueux replied:
no, it isn't a error.

autra replied:
Then `return EMPTY_TEXTURE_ZOOM`? My point is that we shouldn't return a magic number.

peppsac replied:
Comment not addressed.

autra commented on src/Core/Commander/Scheduler.js:
```diff
@@ -120,7 +120,10 @@ Scheduler.prototype.execute = function execute(command) {
 
     // execute command now if possible
-    if (q.counters.executing < this.maxCommandsPerHost) {
+    if (forceRunNow) {
```
Not sure about this. Wouldn't it be simpler to simply call `provider.executeCommand` if `searchInParent` is true in NodeProcess.processNodeImagery or Elevation?

autra reviewed (changes requested):
One more step and I guess we'll be good :-)

peppsac commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -109,147 +86,69 @@ WMTS_Provider.prototype.url = function url(coWMTS, layer) {
+                parentTextures[0].image.data,
+                SIZE_TEXTURE_TILE, SIZE_TEXTURE_TILE,
+                pitch);
+        return Promise.resolve({ pitch, texture: parentTextures[0], min, max });
```
I'd suggest to *not* use `Promise` where not needed.
Just return your object if this task can be performed directly.
You'll to adapt `Scheduler` line:
```
var p = provider.executeCommand(cmd) || Promise.resolve();
```
to test if the result from `executeCommand` is a Promise or a plain object.

peppsac commented on src/Core/Commander/Scheduler.js:
```diff
@@ -101,7 +101,7 @@ Scheduler.prototype.runCommand = function runCommand(command, queue, executingCo
 };
 
-Scheduler.prototype.execute = function execute(command) {
+Scheduler.prototype.execute = function execute(command, forceRunNow) {
```
Rather than adding a boolean which makes code reading harder, I'd prefer a 2nd function.
```js
sheduler.execute(cmd, false);
sheduler.execute(cmd, true);
```
is less clear than:
```js
sheduler.execute(cmd);
sheduler.executeNow(cmd);
```
You could even change the return type for each function.
`execute` would return a Promise, while `executeNow` would return the result (and fail with an error if the requested command cannot be executed in a synchronous fashion).

autra replied:
I'm in favor of this change. let's not add more complexity to `execute`

peppsac replied:
`forceRunNow` option is never used now (which is good), so please remove it.

peppsac commented:
Maybe we could simplify the process of inheriting its parent's textures for a node.

If texture had another attribute describing the tile area covered, we could deduce the offset and scale to use for a child without asking the original provider (and it's a trivial function to write).

With this change, we wouldn't even need to add a `forceRunNow` option to the scheduler.

Does this make sense or am I missing something?

autra commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -109,147 +86,69 @@ WMTS_Provider.prototype.url = function url(coWMTS, layer) {
+    const pitch = new THREE.Vector3(0.0, 0.0, 1.0);
+
+    if (parentTextures) {
+        WMTS_WGS84Parent(
```
Well all right. I see. This WMTS_WGS84Parent function is really bad. It has side-effects AND returns a value. This is very confusing and error-prone. I'm not requiring anything here as it is just a copy-paste, but we might want to refacto that in the future.

autra reviewed (dismissed):
Still https://github.com/iTowns/itowns2/pull/244/files#r111940993 and last @peppsac comment to address/answer and we will be done here I guess :-)

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -76,7 +90,7 @@ var moveElementsArraySafe = function moveElementsArraySafe(array,index, howMany,
 /* eslint-enable */
 
-const LayeredMaterial = function LayeredMaterial(id) {
+const LayeredMaterial = function LayeredMaterial(material, coord) {
```
The rest look good so I'll focus only on this part.

Here's what I had in mind:

- Replace LayeredMaterial `offsetScale` with:
```js
// express what part of the texture (source) is drawn,
// and where (destination) it is drawn.
// (mapping.source is our current `offsetScale`)
// (mapping.destination is currently baked into UV attributes of a tile)
mapping = {
    source: {
        origin: new THREE.Vector2(0, 0),
        size: 1,
    },
    destination: {
        origin: new THREE.Vector2(0, 0),
        size: 1,
    },
}
```

- replace `pitch` result from providers with a `mapping` object. Currently
  only PM textures will have a non-default destination

- replace `coord` in `function LayeredMaterial(material, coord)` with tile's mapping attribute,
  as well as it's parent's mapping attribute.
  With this change, `offsetToParent` can be written without any dependencies on WMTS logic.

What do you think ?

gchoqueux replied:
I would like to do. We need to talk about this.

-  `mapping` will not bear directly into uniforms

gchoqueux replied:
> dependencies on WMTS logic.

```js
var emptyTexture = new THREE.Texture();
emptyTexture.coordWMTS = { zoom: EMPTY_TEXTURE_ZOOM };
```

There was already a dependency with WMTS. But it should also be deleted

gchoqueux replied:
As some projections are not linear (like PM), it is simpler to do the calculations in the same system and use WMTS coordinate.

gchoqueux replied:
I think to keep WMTS logic.
it's more simple to work in same system (specially for non-linear systems)
What do you think?

peppsac replied:
I disagree :)
But that something we can fix later, so I won't block any longer on this point.

peppsac commented on src/Scene/Quadtree.js:
```diff
@@ -24,7 +24,7 @@ function Quadtree(type, schemeTile, link) {
     this.tileType = type;
     this.minLevel = 2;
-    this.maxLevel = 17;
+    this.maxLevel = 18;
```
This change is unrelated to the PR (as is the removal of imagery layers).
You should be extra careful to not mix these kind of changes in a refactoring PR.

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -146,11 +160,56 @@ const LayeredMaterial = function LayeredMaterial(id) {
+    this.elevationLayersId = [];
+
+    if (material instanceof LayeredMaterial) {
+        // Getting inheritance from elevation
```
May I suggest this comment instead:
```js
// inherit parent's elevation texture
```
(and same for the similar ones below)

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -146,11 +160,56 @@ const LayeredMaterial = function LayeredMaterial(id) {
+
+    if (material instanceof LayeredMaterial) {
+        // Getting inheritance from elevation
+        if (material.getElevationLayerLevel() > -1) {
```
`!= EMPTY_TEXTURE_ZOOM` ?

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -146,11 +160,56 @@ const LayeredMaterial = function LayeredMaterial(id) {
+        let offsetTextures = 0;
+        for (let i = 0; i < material.colorLayersId.length; i++) {
+            const layerId = material.colorLayersId[i];
+            if (material.getColorLayerLevelById(layerId) > -1) {
```
`!= EMPTY_TEXTURE_ZOOM` ?

peppsac replied:
The whole loop could benefit from a comment explaining its purpose (and "how" it is achieved).

peppsac reviewed (changes requested):
A few comments to address (please make sure you reply to all of them, so reviewers know that you took a look at them).
As wrote in another comment, this is my last "Request changes" review for this PR - next time I'll "Approve" it so we can move forward.

peppsac commented on src/Core/Commander/Providers/OGCWebServiceHelper.js:
```diff
@@ -37,43 +38,50 @@ export default {
             return result;
         });
+
+        cachePending[url] = promiseXBil;
```
`cachePending` will grow indefinitely.
You could probably remove elements from it after the line:
```js
cache.addRessource(url, result.texture);
```

peppsac commented on src/Globe/Globe.js:
```diff
@@ -202,6 +202,16 @@ Globe.prototype.removeColorLayer = function removeColorLayer(layer) {
+Globe.prototype.update = function update() {
+    var cO = function cO(object) {
+        if (object.loaded !== undefined) {
+            object.loaded = false;
```
What is the purpose of this function?

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -187,17 +158,25 @@ NodeProcess.prototype.hideNodeChildren = function hideNodeChildren(node) {
-    } else if (node.isDisplayed()) {
+
+
+    // NOTE : node.loaded is removed
```
Remove this - no need to keep comments refering to removed code

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -187,17 +158,25 @@ NodeProcess.prototype.hideNodeChildren = function hideNodeChildren(node) {
+    // Priorisation of the most degraded tiles
+    const priority = node.level * Math.round(node.sse);
+
+    if (node.isDisplayed()) {
```
Again: this is a functionnal change in a refacto PR.
Avoid this please.

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -225,6 +205,25 @@ function updateNodeImagery(scene, quadtree, node, layersConfig, force) {
+        const material = node.materials[RendererConstant.FINAL];
+
+
+        if (material.indexOfColorLayer(layer.id) === -1) {
```
`=== EMPTY_TEXTURE_ZOOM`?

gchoqueux replied:
is index not level

gchoqueux commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -76,7 +90,7 @@ var moveElementsArraySafe = function moveElementsArraySafe(array,index, howMany,
 /* eslint-enable */
 
-const LayeredMaterial = function LayeredMaterial(id) {
+const LayeredMaterial = function LayeredMaterial(material, coordDestination, coordSource) {
```
@peppsac , you disagree, me too I prefer a most generic thing, as you suggest.
I ask for another opinion because I think it isn't possible with this `mapping` method,  due to nonlinear projections (PM, Lambert...) This method requires several unnecessary conversions. On the other hand it is perfect in the WGS84 system
```js
mapping = {
    source: {
        origin: new THREE.Vector2(0, 0),
        size: 1,
    },
    destination: {
        origin: new THREE.Vector2(0, 0),
        size: 1,
    },
}
```

gchoqueux committed:
refacto (nodeProcess, layeredMaterial, tileMesh)

* LayeredMaterial : new instance can inherit from parent

* NodeProcess :
    - Move initialisation Imagery layer in updateNodeImagery
    - Move initialisation elevation layer in updateNodeElevation

* tileMesh : remove loaded statut and texturesNeeded due to the new heritage of LayeredMaterial

* Fix concurrence loading elevation due to the new heritage of LayeredMaterial

peppsac commented on src/Scene/BoundingBox.js:
```diff
@@ -138,9 +138,16 @@ BoundingBox.prototype.offsetScale = function offsetScale(bbox) {
+        y: Math.abs(this.north() - this.south()),
+    };
+
+    var pitX = Math.abs(bbox.west() - this.west()) / dimension.x;
```
I don't think you want to use `Math.abs` here.
Instead you should check that inputs are valid (ie : that `bbox` is contained in `this`) and throw an error otherwise

peppsac replied:
Also I'd prefer `originX` and `originY` to `pitX` and `pitY`

peppsac commented on src/Core/Commander/Providers/OGCWebServiceHelper.js:
```diff
@@ -0,0 +1,92 @@
+// The PM textures overlap several tiles WGS84, it is to avoid net requests
+// Info : THREE.js have cache image https://github.com/mrdoob/three.js/blob/master/src/loaders/ImageLoader.js#L25
+const cache = CacheRessource();
+const cachePending = [];
```
How about using a [Map](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Map) instead?

peppsac reviewed (commented):
Mostly good, just a few things to address and we're good to go
